### PR TITLE
Define GenerateConfig interface and its sub-interfaces

### PR DESCRIPTION
**Pull request by oliversun9**
This PR defines `GenerateConfig` interface and its sub-interfaces.
More spefically
- GeneratePluginConfig. This is taken mostly from work in `bufgenv2`, where two versions of configs are already sharing the same plugin config interface.
- GenerateManagedConfig. I anticipate discussion for this one. I had some alternatives defined [in a scratch space](https://github.com/bufbuild/buf/compare/bufmod...osun/gen-config-stuff#diff-44b7f92a514fc52e6785fe3d4624b51a16aa7105c7529d1f12ed49f193c964eaR39) with how callers might use them.
- GenerateTypeConfig. Very straightforward, only v1 needs it.
- GenerateInputConfig. It's left as TODO for v2, but its shape might very much look like what's in `bufgenv2`

### CODE REVIEW

File: private/bufnew/bufconfig/generate_managed_config.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	// 2. In the config the user has specified not to modify this option for this file.
+	// 3. The file already has the desired file option value.
+	// 4. The file is a WKT.
+	// TODO: 3 and 4 are debatable, asking the question what is the responsibility
+	// of a managed config.
+	//
+	// Note: this means a GenerateManagedConfig's interface does not deal with
+	// options like java_package_prefix and java_package_suffix, because it returns
+	// the final value to modify java_package to.
+	ValueForFileOption(imageFile, fileOption) (interface{}, bool)
```
Comment (bufdev) [inline private/bufnew/bufconfig/generate_managed_config.go:42]: How am I supposed to use this outside of this package?

File: private/bufnew/bufconfig/generate_config.go [hunk 2]
```diff
@@ -15,47 +15,36 @@
...
 
 // GenerateConfig is a generation configuration.
-//
-// TODO
 type GenerateConfig interface {
+	// GeneratePluginConfigs returns the plugin configurations. This will always be
+	// non-empty. Zero plugin configs will cause an error at construction time.
 	GeneratePluginConfigs() []GeneratePluginConfig
-	// may be nil
+	// GenerateManagedConfig returns the managed mode configuration. This is never nil.
```
Comment (bufdev) [inline private/bufnew/bufconfig/generate_config.go:22]: Why are we taking the position of not using `nil` as an indicator of non-existence here, but using it with `GenerateTypeConfig`?

File: private/bufnew/bufconfig/generate_plugin_config.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	// This is the default value for local plugins.
+	GenerateStrategyDirectory GenerateStrategy = 1
+	// GenerateStrategyAll is the strategy to generate with all files at once.
+	//
+	// This is the only strategy for remote plugins.
+	GenerateStrategyAll GenerateStrategy = 2
+)
+
+// GeneratePluginConfig is a configuration for a plugin.
+type GeneratePluginConfig interface {
```
Comment (bufdev) [inline private/bufnew/bufconfig/generate_plugin_config.go:33]: Are you asking callers to do a type switch to figure out what type of interface they are using?

File: private/bufnew/bufconfig/generate_managed_config.go (hunk 1 above)
Comment (oliversun9) [reply private/bufnew/bufconfig/generate_managed_config.go:42]:
```
for _, imageFile := range imageFiles {
  javaPackage, shouldModify := config.ValueForFileOption(imageFile, fileOptionJavaPackage)
  javaPackageValue, ok := javaPackage.(string)
  if !ok {
    return someErr
  }
  if !shouldModify {
    continue
  }
  imageModify.ModifyJavaPackage(marker, imageFile, javaPackageValue)
  
  // continue to the next option
}
```

or perhaps something like
```
        for _, imageFile := range imageFiles {
		for _, fileOption := range fileOptions {
			getFunc := option.GetFunc() //
			setFunc := option.SetFunc()
			value, shouldModify := config.ValueForFileOption(option)
			if !set {
				continue
			}
			if value == getFunc(imageFile) {
				continue
			}
			setFunc(marker, imageFile, value)
		}
		for _, field := range image.Fields := {
			for _, fieldOption := range fieldOptions {
			}
		}
	}
```

**Review by bufdev — COMMENTED** [review]
Overall looks really good - we can do a line by line tomorrow, but I'd say you're good to start implementing, the overall shape looks correct

File: private/bufnew/bufconfig/generate_managed_config.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+// i.e. At least one of Path, ModuleFullNameString, FieldName, FileOption and
+// FieldOption is not empty. A rule can disable all options for certain files/fields,
+// disable certains options for all files/fields, or disable certain options for
+// certain files/fields. To disable all options for all files/fields, turn off managed mode.
+type ManagedDisableRule interface {
+	// Path returns the file path, relative to its module, to disable managed mode for.
+	Path() string
+	// ModuleFullNameString returns the full name string of the module to disable
+	// managed mode for.
+	ModuleFullNameString() string
```
[the remark this answers is not in the document]
Comment (oliversun9) [reply private/bufnew/bufconfig/generate_managed_config.go]: fixed in [0715e66](https://github.com/bufbuild/buf/pull/2617/commits/0715e66b57ed8650e6f38998a4afe7a659c1a12e)

File: private/bufnew/bufconfig/generate_managed_config.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+//   - The files/fields for which the options are modified. If all of Path, ModuleFullNameString
+//   - or FieldName are empty, all files/fields are modified. Otherwise, only
+//     file/fields that match the specified Path, ModuleFullNameString and FieldName
+//     is modified.
+type ManagedOverrideRule interface {
+	// Path is the file path, relative to its module, to disable managed mode for.
+	Path() string
+	// ModuleFullNameString is the full name string of the module to disable
+	// managed mode for.
+	ModuleFullNameString() string
```
[the remark this answers is not in the document]
Comment (oliversun9) [reply private/bufnew/bufconfig/generate_managed_config.go]: fixed in [0715e66](https://github.com/bufbuild/buf/pull/2617/commits/0715e66b57ed8650e6f38998a4afe7a659c1a12e)

File: private/bufnew/bufconfig/generate_managed_config.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	Prefix() string
+	// Suffix returns the override suffix.
+	Suffix() string
+
+	isManagedOverride()
+}
+
+// bufimage.ImageFile implements this, but this also allows private implementation
+// for testing.
+type ImageFile interface {
```
[the remark this answers is not in the document]
Comment (oliversun9) [reply private/bufnew/bufconfig/generate_managed_config.go]: removed in [7b50b79](https://github.com/bufbuild/buf/pull/2617/commits/7b50b7943c21aac7a2fb97092325e22748b7bf44)

File: private/bufnew/bufconfig/generate_managed_config.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	// Path returns the path relative to the its module's root.
+	Path() string
+	// FileDescriptorProto returns the its field descriptor.
+	FileDescriptorProto() *descriptorpb.FileDescriptorProto
+}
+
+// FileOption is a file option.
+type FileOption int
+
+const (
```
[the remark this answers is not in the document]
Comment (oliversun9) [reply private/bufnew/bufconfig/generate_managed_config.go]: fixed in [99b3741](https://github.com/bufbuild/buf/pull/2617/commits/99b374156a497ef13e2e63b720d1bd3bb89b4e69)

**Pull request closed without merging**